dask/dataframe/io.py

```diff
@@ -406,65 +407,96 @@ def _link(token, result):
 
 @wraps(pd.DataFrame.to_hdf)
 def to_hdf(df, path_or_buf, key, mode='a', append=False, complevel=0,
-           complib=None, fletcher32=False, get=get_sync, dask_kwargs=None,
-           name_function=None, compute=True, **kwargs):
+           complib=None, fletcher32=False, get=None, dask_kwargs={},
+           name_function=None, compute=True, lock=None, **kwargs):
     name = 'to-hdf-' + uuid.uuid1().hex
 
     pd_to_hdf = getattr(df._partition_type, 'to_hdf')
 
+    single_file = True
+    single_node = True
+
     # if path_or_buf is string, format using i_name
     if isinstance(path_or_buf, str):
         if path_or_buf.count('*') + key.count('*') > 1:
             raise ValueError("A maximum of one asterisk is accepted in file path and dataset key")
 
         fmt_obj = lambda path_or_buf, i_name: path_or_buf.replace('*', i_name)
+
+        if '*' in path_or_buf:
+            single_file = False
     else:
         if key.count('*') > 1:
             raise ValueError("A maximum of one asterisk is accepted in dataset key")
 
         fmt_obj = lambda path_or_buf, _: path_or_buf
 
+    if '*' in key:
+        single_node = False
+
     if name_function is None:
         name_function = build_name_function(df.npartitions - 1)
 
     # we guarantee partition order is preserved when its saved and read
     # so we enforce name_function to maintain the order of its input.
-    if '*' in key or (isinstance(path_or_buf, str) and '*' in path_or_buf):
+    if not (single_file and single_node):
         formatted_names = [name_function(i) for i in range(df.npartitions)]
         if formatted_names != sorted(formatted_names):
-            warn("In order to preserve order between partitions "
-                 "name_function must preserve the order of its input")
+            warn("To preserve order between partitions name_function "
+                 "must preserve the order of its input")
 
+    # when writing to the same file in multiple tasks we should lock
+    if lock is None:
+        if not single_node:
+            lock = True
+        else:
+            lock = False
+
+    # if get not specified default to sync if we're writing to a single entity
+    # if we're writing to multiple files, default to multiprocessing
+    # by default hdf5 is not thread safe, so threaded is not an option
+    if get is None:
+        if not single_file:
+            get = dask.multiprocessing.get
+        else:
+            get = get_sync
```

Comment:
I propose that the two options here should be "synchronous if everything is linked, or whatever the default was before otherwise".  This is generally done by `get=None`